src/io/flutter/vmService/VMServiceManager.java

```diff
@@ -560,6 +578,120 @@ private void unregisterService(String serviceName) {
+   * <p>
+   * The current value of the subscription can be null occasionally during initial application startup and for a brief time when doing a
+   * hot restart.
+   */
+  public StreamSubscription<Double> getCurrentDisplayRefreshRate(Consumer<Double> onValue, boolean onUIThread) {
+    return displayRefreshRateStream.listen(onValue, onUIThread);
+  }
+
+  /**
+   * Return the current display refresh rate, if any.
+   * <p>
+   * Note that this may not be immediately populated at app startup for Flutter apps. In that case, this will return
+   * the default value (defaultRefreshRate). Clients that wish to be notified when the refresh rate is discovered
+   * should prefer the StreamSubscription variant of this method (getCurrentDisplayRefreshRate()).
+   */
+  public Double getCurrentDisplayRefreshRateRaw() {
+    synchronized (displayRefreshRateStream) {
+      Double fps = displayRefreshRateStream.getValue();
+      if (fps == null) {
+        fps = defaultRefreshRate;
+      }
+      return fps;
+    }
+  }
+
+  public CompletableFuture<Double> getDisplayRefreshRate() {
+    final double unknownRefreshRate = 0.0;
+
+    final String flutterViewId = getFlutterViewId().exceptionally(exception -> {
+      LOG.warn(exception.getMessage());
+      return null;
+    }).join();
+
+    if (flutterViewId == null) {
+      // Fail gracefully by returning the default.
+      return CompletableFuture.completedFuture(defaultRefreshRate);
+    }
+
+    return invokeGetDisplayRefreshRate(flutterViewId);
+  }
+
+  private CompletableFuture<Double> invokeGetDisplayRefreshRate(String flutterViewId) {
+    final CompletableFuture<Double> ret = new CompletableFuture<>();
+    final JsonObject params = new JsonObject();
+    params.addProperty("viewId", flutterViewId);
+    vmService.callServiceExtension(
+      getCurrentFlutterIsolateRaw().getId(), ServiceExtensions.displayRefreshRate, params,
+      new ServiceExtensionConsumer() {
+        @Override
+        public void onError(RPCError error) {
+          ret.completeExceptionally(new RuntimeException(error.getMessage()));
+        }
+
+        @Override
+        public void received(JsonObject object) {
+          if (object == null) {
+            ret.complete(null);
+          }
+          else {
+            ret.complete(object.get("fps").getAsDouble());
+          }
+        }
+      }
+    );
+    return ret;
+  }
+
+  private CompletableFuture<String> getFlutterViewId() {
+    return getFlutterViewsList().thenApplyAsync((JsonElement element) -> {
+      final JsonArray viewsList = element.getAsJsonObject().get("views").getAsJsonArray();
+      for (JsonElement jsonElement : viewsList) {
+        final JsonObject view = jsonElement.getAsJsonObject();
+        if (view.get("type").getAsString().equals("FlutterView")) {
+          return view.get("id").getAsString();
+        }
+      }
+      throw new RuntimeException("No Flutter views to query: " + viewsList.toString());
```

Comment:
will this exception get surfaced to users? it could make sense to fail gracefully if flutter happens to not really be running any views yet.

Reply:
No it won't. We will log it as a warning and then return the default refresh rate.